src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -136,28 +137,68 @@ public ItemData(I item, int elementOrder, bool conditionResult)
         private class OperationWithCache : IItemOperation
         {
             private readonly IItemOperation _operation;
-            private readonly Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>> _cache;
+            private Dictionary<ISet<string>, ImmutableList<ItemData>> _cache;
+
+            private bool _isReferenced;
+#if DEBUG
+            private bool _applyCalled;
+#endif
 
             public OperationWithCache(IItemOperation operation)
             {
                 _operation = operation;
-                _cache = new Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>>(
-                    () => new Dictionary<ISet<string>, ImmutableList<ItemData>>());
             }
 
             public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+#if DEBUG
+                Debug.Assert(!_applyCalled, "Apply should only be called once. Otherwise caching is not working");
+#endif
+
                 _operation.Apply(listBuilder, globsToIgnore);
+
+                // cache results if somebody is referencing this operation
+                if (_isReferenced)
+                {
+                    AddItemsToCache(globsToIgnore, listBuilder.ToImmutable());
+                }
+#if DEBUG
+                _applyCalled = true;
+#endif
             }
 
             public bool TryGetFromCache(ISet<string> globsToIgnore, out ImmutableList<ItemData> items)
             {
-                return _cache.Value.TryGetValue(globsToIgnore, out items);
+                if (_cache != null)
+                {
+                    return _cache.TryGetValue(globsToIgnore, out items);
+                }
+
+                items = null;
+                return false;
+            }
+
+            /// <summary>
+            /// Somebody is referencing this operation
+            /// </summary>
+            public void AddReference()
```

Comment:
It might be better to find a different name.  `AddReference` sounds to me like this is doing ref counting.